src/Raven.Server/Documents/Indexes/LastProcessedReference.cs

```diff
@@ -0,0 +1,62 @@
+﻿using System.Collections.Generic;
+using Raven.Server.Documents.Indexes.Workers;
+
+namespace Raven.Server.Documents.Indexes
+{
+    public class LastProcessedReference
+    {
+        public enum ActionType
+        {
+            Document,
+            Tombstone
+        }
+        private class ReferenceState
+        {
+            public string ReferencedDocumentId;
+            public string NextItemId;
+            public long ReferenceDocumentEtag;
+        }
+        
+        private readonly Dictionary<string, ReferenceState> _lastIdPerCollectionForDocuments = new Dictionary<string, ReferenceState>();
+
+        private readonly Dictionary<string, ReferenceState> _lastIdPerCollectionForTombstones = new Dictionary<string, ReferenceState>();
+
+        public void Set(ActionType actionType, string collection, HandleReferencesBase.Reference referencedDocument, string itemId)
```

Comment:
referencedDocument => reference (it is not always document)

Reply:
done